Sources/BraintreeCore/BTAPIClient.swift

```diff
@@ -93,23 +93,20 @@ import Foundation
     /// cached on subsequent calls for better performance.
     @_documentation(visibility: private)
     public func fetchOrReturnRemoteConfiguration(_ completion: @escaping (BTConfiguration?, Error?) -> Void) {
-        configurationLoader.getConfig { [weak self] configuration, error in
-            guard let self else {
-                completion(nil, BTAPIClientError.deallocated)
-                return
-            }
-            
-            if let error {
+        // TODO: - Consider updating all feature clients to use async version of this method?
+
+        Task { @MainActor in
```

Comment:
Just making sure that these don't cause deadlocks with nested calls to fetchOrReturnConfig.

Reply:
`@MainActor` is equivalent to running on the main dispatch queue: https://developer.apple.com/documentation/swift/mainactor. So should behave the exact same.

Reply:
I think main difference for our purposes is that @MainActor can have suspension points so we wouldn't have deadlocks with nested completion handlers with fetchOrReturnRemoteConfiguration

Reply:
I'm just puzzled why there are differences in analytics output with this[ commit](20d7a46d9381de1d6e01fd601c52f14fa668dbf5).

Reply:
I guess we are just doing same on two call sites, so coordinating the two functions calls.

Reply:
Yep exactly, we just want them to both behave the same, will also make it easier to move to the async method when we tackle that ticket.

Reply:
the ConfigurationLoader's getConfig method is annotated with globalActor, so I think they should be serialized.

Reply:
Will this be a blocking action on the main thread? If so, that would interrupt a merchant doing UI updates

Reply:
cc: @jaxdesmarais @KunJeongPark

Reply:
I don't think it's blocking, because MainActor allows suspension points and Configuration.getConfiguration runs on its own actor, actual network call is not performed on main thread.

I think there might be unnecessary thread hopping.

Reply:
MainActor should be reserved for UI tasks only ([doc](https://www.avanderlee.com/swift/mainactor-dispatch-main-thread/)). Why was this needed at all?

Reply:
@scannillo from that doc: `Dispatching to the main actor only takes place if needed to ensure the best performance possible.`. It is worth noting this has always taken place on the main thread, see the following test: https://github.com/braintree/braintree_ios/blob/c17642dad8f8a45f1b2d51c0c762873d35e17f6e/UnitTests/BraintreeCoreTests/BTAPIClient_Tests.swift#L264-L268

We may be able to clean this up slightly, but this is not new behavior to have `fetchOrReturnRemoteConfiguration` run on main.

Reply:
Digging further into this and the `@MainActor` for the async `fetchConfiguration` we were not properly caching the config and pulling from the cached config if not using `@MainActor`, additionally removing this causes the test `testGetConfig_whenCalledInQuickSequence_onlySendsOneNetworkRequest` to fail as we are making 4 network request. We expect to be doing these calls off of the main thread (again the main thread is an expectation in the test above). I am going to leave these as is to ensure our current behavior based on our tests and my observations from FPTI.

Reply:
That test` testGetConfig_whenCalledInQuickSequence_onlySendsOneNetworkRequest` just runs on the function on ConfigurationLoader `getConfig`. I'm not sure how removing MainActor annotation on the function that calls it, fetchConfiguration affects that test. I've shaven't seen that test fail on its own.

Reply:
That test you quoted above,` testCallbacks_useMainDispatchQueue()`, it expects the completion handler to return on main.
So entire Task block of  fetchOrReturnRemoteConfiguration doesn't need to be annoytated @MainActor,
just the setupHTTPCredentials and completion invocation need to be wrapped in await MainActor.run to make this
test pass.

Reply:
I haven't personally seen weirdness or difference in the FPTI between what was on main before, before this crash test was implemented, after adding it, with or without MainActor annotations for fetchConfig functions.

But I trust that you saw something, there is a lot going on with sending analytics. 
Let's keep an eye on latency metrics and then we can revisit. What are your thoughts on that?

Reply:
> That test you quoted above, testCallbacks_useMainDispatchQueue(), it expects the completion handler to return on main.
So entire Task block of fetchOrReturnRemoteConfiguration doesn't need to be annotated @mainactor,
just the setupHTTPCredentials and completion invocation need to be wrapped in await MainActor.run to make this test pass, as I mentioned at top of this thread.

imo, adding `await MainActor.run` for just the completions/setupHTTPCredentials complicates the callsite vs running the task on `MainActor`. There is no observed latency since `getConfig` is a global isolated actor already. Having a cleaner callsite is more important as there is no difference in wrapping the completions in a `MainActor.run` vs what exists today, but this method will also entirely be removed in a future PR.

> I haven't personally seen weirdness or difference in the FPTI between what was on main before, before this crash fix was implemented, after adding it, with or without MainActor annotations for fetchConfig functions.

If  you look at ticket DTMOBILES-998 you will see this same behavior where we make multiple network called and the config isn't cached properly. There are screenshots in that ticket with additional details from FPTI.

> My main concerns are potential for more latency for fetching configuration for sendQueuedAnalyticsEvents with @mainactor annotation. But it's hard to tell. Let's keep an eye on latency metrics and then we can revisit.

There is clearly a need to annotate as is, removing it causes the issues we can clearly see in FPTI noted above. Removing the duplicate config call and just relying on the async method should resolve any remaining confusion. There is not additional latency added that did not exist previously.

Reply:
Thank you for the screenshots. Do these extra calls happen in 6.23.2? I can check.
I believe you that you observed these, I just didn't see it myself. I am trying to figure out why.
I want to make sure I understand what @MainActor annotation is fixing.

Reply:
I made comments on your jira ticket with screenshots. I tested 6.23.3 with @mainactor annotation as it is currently on main wit no changes, 6.23.3 without @mainactor annotation but with await MainActor.run on setHTTPCredentials and completion calls and 6.23.2 with no changes.

I didn't incorporate your changes on this branch, I am going to try that as well.

I found that there are no duplicate calls to v1/configuration in any of them, they are all fetching from cache but 6.23.3 without @mainactor annotation results in analytic events being listed out of order, this was also true for 6.23.2, the analytics were out of order.
I just ran through each once, so I will try several times for verification.